src/debug_tools/diff.rs

```diff
@@ -0,0 +1,507 @@
+    // there is a hash diff deep in two separate trie structures doesn't make much
+    // sense).
+    longest_state
+        .longest_key_node_diff
+        .or(longest_state.longest_key_hash_diff)
+}
+
+#[derive(Debug, Default)]
+struct DepthNodeDiffState {
+    longest_key_node_diff: Option<DiffPoint>,
+    longest_key_hash_diff: Option<DiffPoint>,
+}
+
+impl DepthNodeDiffState {
+    fn try_update_longest_divergence_key_hash(&mut self, state: &DepthDiffPerCallState) {
+        Self::replace_longest_field_if_our_key_is_larger(
+            &mut self.longest_key_hash_diff,
+            &state.curr_key,
+            state.a,
+            state.b,
+            state.curr_path.clone(),
+        );
+    }
+
+    fn try_update_longest_divergence_key_node(&mut self, state: &DepthDiffPerCallState) {
+        Self::replace_longest_field_if_our_key_is_larger(
+            &mut self.longest_key_node_diff,
+            &state.curr_key,
+            state.a,
+            state.b,
+            state.curr_path.clone(),
+        );
+    }
+
+    fn replace_longest_field_if_our_key_is_larger(
+        field: &mut Option<DiffPoint>,
+        parent_k: &Nibbles,
+        child_a: &HashedPartialTrie,
+        child_b: &HashedPartialTrie,
+        path: NodePath,
+    ) {
+        if field
+            .as_ref()
+            .map_or(true, |d_point| d_point.key.count < parent_k.count)
+        {
+            *field = Some(DiffPoint::new(child_a, child_b, *parent_k, path))
+        }
+    }
+}
+
+/// State that is copied per recursive call.
+#[derive(Clone, Debug)]
+struct DepthDiffPerCallState<'a> {
+    a: &'a HashedPartialTrie,
+    b: &'a HashedPartialTrie,
+    curr_key: Nibbles,
+    curr_depth: usize,
+
+    // Horribly inefficient, but these are debug tools, so I think we get a pass.
+    curr_path: NodePath,
+}
+
+impl<'a> DepthDiffPerCallState<'a> {
+    /// Exists solely to prevent construction of this type from going over
+    /// multiple lines.
+    fn new(
+        a: &'a HashedPartialTrie,
+        b: &'a HashedPartialTrie,
+        curr_key: Nibbles,
+        curr_depth: usize,
+    ) -> Self {
+        Self {
+            a,
+            b,
+            curr_key,
+            curr_depth,
+            curr_path: NodePath::default(),
+        }
+    }
+
+    /// Note: The assumption here is that `a` and `b` are of the same node type
+    /// and have the key.
+    fn new_from_parent(
+        &self,
+        a: &'a HashedPartialTrie,
+        b: &'a HashedPartialTrie,
+        key_piece: &Nibbles,
+    ) -> Self {
+        let new_segment = get_segment_from_node_and_key_piece(self.a, key_piece);
+        let new_path = self.curr_path.dup_and_append(new_segment);
+
+        Self {
+            a,
+            b,
+            curr_key: self.curr_key.merge_nibbles(key_piece),
+            curr_depth: self.curr_depth + 1,
+            curr_path: new_path,
+        }
+    }
+}
+
+fn find_diff_point_where_tries_begin_to_diff_depth_rec(
```

Comment:
We may want to rename it to something more compact / meaningful?

Reply:
Yep!